Systemdtest

jan-cerny opened the pull request:
This pull request contains new checks for Fedora Content. They check if auditd and chronyd services are enabled on the system. The new systemdunitdependency_test, systemdunitdependency_object and systemdunitdependency_state are used. Those elements were introduced in OVAL 5.11. Therefore you need OpenSCAP supporting OVAL 5.11 to make the new content. OVAL 5.11 is supported since OpenSCAP 1.2.2. This version is available from a COPR repository https://copr.fedoraproject.org/coprs/isimluk/OpenSCAP 
The user is informed by a message if he tries to make the content with older versions of OpenSCAP and the target is skipped.

The ntpd was replaced by chronyd because chronyd in installed by default on Fedora. The duplicate check  in Fedora/input/system/auditing.xml was removed due to conflict with new service_auditd_enabled check.

Please see https://lists.fedorahosted.org/pipermail/scap-security-guide/2015-April/006405.html

redhatrises commented on Fedora/input/checks/chronyd_specify_remote_server.xml:
```diff
@@ -0,0 +1,28 @@
+    <metadata>
+      <title>Specify a Remote NTP Server for Time Data</title>
+      <affected family="unix">
+        <platform>multi_platform_rhel</platform>
```
This `<platform>` check should not exist unless this check is for both RHEL 6&7 and Fedora and should be in `shared/oval` if this check is RHEL is supported.

redhatrises commented on Fedora/input/checks/chronyd_specify_multiple_servers.xml:
```diff
@@ -0,0 +1,28 @@
+    <metadata>
+      <title>Specify a Remote NTP Server for Time Data</title>
+      <affected family="unix">
+        <platform>multi_platform_rhel</platform>
```
This `<platform>` check should not exist unless this check is for both RHEL 6&7 and Fedora and should be in `shared/oval` if this check is RHEL is supported.

redhatrises commented on config/oval.config:
```diff
@@ -31,4 +31,4 @@ multi_platform_rhev =
 #
 [Schema Version]
-oval_version = 5.10
+oval_version = 5.11
```
Is there a supported `openscap-scanner` that works with OVAL 5.11 on RHEL 6 & 7?

isimluk replied:
No there is not.

I have brought this issue on the mailing list for wider awareness. https://lists.fedorahosted.org/pipermail/scap-security-guide/2015-April/006405.html

redhatrises commented on shared/modules/testcheck_module.py:
```diff
@@ -23,7 +23,7 @@
         <oval:product_name>testcheck.py</oval:product_name>
         <oval:product_version>0.0.1</oval:product_version>
-        <oval:schema_version>5.10</oval:schema_version>
+        <oval:schema_version>5.11</oval:schema_version>
```
Is 5.11 supported on RHEL 6&7?

jan-cerny replied:
I realized that this must be done differently. This would affect also the RHEL content. It should be re-factored to influence only the Fedora content.

redhatrises commented on shared/utils/verify-input-sanity.py:
```diff
@@ -31,7 +31,7 @@
         <oval:product_name>testcheck.py</oval:product_name>
         <oval:product_version>0.0.1</oval:product_version>
-        <oval:schema_version>5.10</oval:schema_version>
+        <oval:schema_version>5.11</oval:schema_version>
```
Is 5.11 supported on RHEL 6&7?

jan-cerny replied:
Does OVAL version matter in this script?

iankko replied:
I would say it matters. The ```verify-input-sanity.py``` checking script should verify particular OVAL check sanity for syntactic correctness. Obviously due to changes in OVAL language 5.10.1 vs 5.11 vs 5.11.1 some elements changed in the sense they aren't inter-version compatible. So would be better RHEL checks to be checked explicitly against 5.10.1 version and introduce own ```verify-input-sanity.py``` copy for Fedora content.

redhatrises commented:
A couple of additional comments:

* If the functionality is desired in `Fedora/compose_ds.sh` and `Fedora/validate.sh`, the code should reside in the `Fedora/Makefile` rather than separate scripts. Also, the OVAL version in `oval.config` should be imported into the `Fedora/Makefile`.

* Also, for Fedora, why is this change tied to a COPR repo? Isn't OpenScap version 1.2.2 already in the Fedora updates repositories?

* What about bumping the required version for Fedora in the `.spec` file?

* These changes do also affect a lot of the RHEL checks as well. Was that also intended?

IMO we should not point to COPR repos.

isimluk commented:
Yeah, we need OVAL 5.11 in Fedora. But we must not broke RHEL6/7 contents.

mpreisler commented:
Don't test for Fedora, test for OVAL 5.11 support in `oscap`. This can be done by grepping through `oscap --v`. You'll probably have to compare oscap versions because of the source datastream situation with OVAL 5.11. The final solution can check for OVAL 5.11 in `oscap --v` and oscap version is not 1.2.1 which AFAIK is the only version that has OVAL 5.11 support but not in datastreams - *check this though, I might be mistaken!*

If there is OVAL 5.11 support there is no reason not to build Fedora content IMO. Even if the user is running RHEL.

Keep in mind that many SSG contributors use RHEL6 or 7. Everything has to build cleanly. The most severe thing I would do is print a warning that Fedora content won't be built because oscap version is too low. Definitely can't cause fatal errors, definitely can't affect RHEL6,7 or the other contents in any way.

Look into `scap-security-guide.spec.in` as well. It has to skip Fedora content if it can't be built on current platform.

iankko commented:
@redhatrises 
> Is there a supported openscap-scanner that works with OVAL 5.11 on RHEL 6 & 7?
Does there need to be? 

What about keeping ```oval_version = 5.10``` in config/oval.config for RHEL, and introducing new ```fedora_oval_version = 5.11``` variable for Fedora (+ editing particular part of ```combinechecks.py``` script it to understand that variable and use OVAL 5.11 for / when generating content for Fedora) ?

redhatrises commented:
@iankko 

> > Is there a supported openscap-scanner that works with OVAL 5.11 on RHEL 6 & 7?

> Does there need to be? 

At some point, at least with RHEL7, yes, there does.

> What about keeping ```oval_version = 5.10``` in config/oval.config for RHEL, and introducing new ```fedora_oval_version = 5.11``` variable for Fedora (+ editing particular part of ```combinechecks.py``` script it to understand that variable and use OVAL 5.11 for / when generating content for Fedora) ?

That should work.

With this PR and others, I am wondering if a build system refactor to handle this case and others would be beneficial as well as more robust?

iankko commented:
@redhatrises 

        Is there a supported openscap-scanner that works with OVAL 5.11 on RHEL 6 & 7?

    Does there need to be?

> At some point, at least with RHEL7, yes, there does.

    What about keeping oval_version = 5.10 in config/oval.config for RHEL, and introducing new fedora_oval_version = 5.11 variable for Fedora (+ editing particular part of combinechecks.py script it to understand that variable and use OVAL 5.11 for / when generating content for Fedora) ?

> That should work. Could also create a config/fedora_oval.config and fix combinechecks.py to handle that.

> With this PR and others, I am wondering if a build system refactor to handle this case and others would be beneficial as well as more robust?

What I was trying to say being that instead of requiring all the SCAP content produced within SSG to have one specific OVAL version, IMHO we should move on and accept the fact different products can have different output / resulting OVAL version. Would the OVAL language promise / follow backward compatibility, when new OVAL language version is available, we would just move all the content to the most recent OVAL language version. But since this is not the case:
  https://oval.mitre.org/language/about/versioning.html

(the major release might break backward compatibility to introduce new features), IMHO the way we should follow is to allow multiple different OVAL language versions to be produced. This will allow us:
* not to need to rewrite old content to new OVAL language version,
* have different products to produce different output OVAL language versions (and simultaneously still have the content in those different products being valid when comes to validating it against concrete OVAL language version),
* last but not least to have content for some product stabilized (in the moment all rules have been implemented && no more bugs have been reported against that content).

Having the OVAL version variable specified in the config file together with the list of products, that OVAL language version should be applied against seems to be more straightforward option (when comes to the need to increase / decrease the set), than to need to modify the build process each time. In other words, it seems more straightforward to me when needing to port some content to different OVAL language version to edit one particular config file row (read as changing:

```-  fedora_oval_version = 5.10```
```+ fedora_oval_version = 5.11```

) than to add additional if / else conditions to the whole build process.

iankko commented:
@redhatrises @jan-cerny 
Was thinking more about how to generate two (or more) different OVAL versions per different products, and the idea is as follows:
* what to introduce "rhel6_oval_version", "rhel7_oval_version", and "fedora_oval_version" variables into ```config/oval.config``` configuration file,
* set values of those variables to expected OVAL versions for that product,
* and update the ```Makefiles``` for each of the products so it would during the build process assign value of that product oval version to main ```oval_version``` variable yet before ```combinechecks.py``` is called (and therefore ```oval_version``` variable value is retrieved) ?

To provide an example for Fedora, the updates could look like as follows:
```
$ git status
# On branch master
# Changes not staged for commit:
#   (use "git add <file>..." to update what will be committed)
#   (use "git checkout -- <file>..." to discard changes in working directory)
#
#	modified:   Fedora/Makefile
#	modified:   config/oval.config
#
no changes added to commit (use "git add" and/or "git commit -a")
```

```
$ git diff config/oval.config
diff --git a/config/oval.config b/config/oval.config
index 386e476..ece7267 100644
--- a/config/oval.config
+++ b/config/oval.config
@@ -27,8 +27,14 @@ multi_platform_rhel = 6,7
 multi_platform_openstack = 
 multi_platform_rhev = 
 
+# Read-only values. DO NOT EDIT
+[OVAL Versions Per Product]
+rhel6_oval_version = 5.10
+rhel7_oval_version = 5.10
+fedora_oval_version = 5.11
+
 #
-# OVAL Schema version
+# Read-write OVAL Schema version. Assign default value
 #
 [Schema Version]
 oval_version = 5.10
```

```
$ git diff Fedora/Makefile
diff --git a/Fedora/Makefile b/Fedora/Makefile
index 1b5614e..363fda3 100644
--- a/Fedora/Makefile
+++ b/Fedora/Makefile
@@ -4,6 +4,7 @@ TRANS = transforms
 SHARED = ../shared
 REFS = $(SHARED)/references
 CONF = ../config
+OVAL_VERSION = $(shell grep "^fedora_oval_version" $(CONF)/oval.config | tr ' ' '\t' | cut -f3)
 UTILS = utils
 DIST = dist
 
@@ -33,6 +34,7 @@ shorthand2xccdf: shorthand-guide
 
 checks:
        xmlwf $(IN)/checks/*.xml
+       sed -i "s/^oval_version = .*/oval_version = $(OVAL_VERSION)/" $(CONF)/oval.config
        $(SHARED)/$(TRANS)/combinechecks.py $(CONF) $(IN)/checks > $(OUT)/unlinked-$(PROD)-oval.xml
        xmllint --format --output $(OUT)/unlinked-$(PROD)-oval.xml $(OUT)/unlinked-$(PROD)-oval.xml
 
```

This will generate OVAL 5.10 content for RHEL-6 and RHEL-7, and 5.11 content for Fedora.

Should the OVAL-5.11 content be desired for RHEL-7 in the future, would be question of simple ```RHEL/7/Makefile``` update too.

Yet via ```fedora_oval_version``` value change in the ```config/oval.config``` file it would be possible to change the generated OVAL version in the future too just for that one concrete product.

iankko commented:
Another option is to update the ```combinechecks.py``` script to:
* determine which product is actually being called on (could be another positional argument from $(PROD) variable),
* based on that $(PROD) it would search ```config/oval.config``` for presence of ```$(PROD)_oval_version``` identifier. If found, use that value,
* if not, default to use ```oval_version``` value as it's doing now

Either version (this or that in the previous comment) works for me.

redhatrises commented:
@iankko I personally like your idea of updating the `combinechecks.py` script as I feel like that would be a better implementation than the Makefile.

jan-cerny commented:
I modified combinechecks.py to have possibility of different OVAL versions for different products. The combinechecks.py has new argument that means name of product. I edited makefiles to pass this argument to the script. I introduced new value "fedora_oval_version" in oval.config.

jan-cerny commented:
@mpreisler : `oscap --version` on OpenSCAP 1.2.1 says that OVAL version is 5.10.1. This means that only version with 5.11 support is OpenSCAP 1.2.2. I think we can simply grep for OVAL version. If it is 5.10.*, we will skip datastream in Fedora content..

iankko commented:
@jan-cerny 
> I modified combinechecks.py to have possibility of different OVAL versions for different products. The combinechecks.py has new argument that means name of product. I edited makefiles to pass this argument to the script. I introduced new value "fedora_oval_version" in oval.config.

Thanks, Jan. There are couple of issues to be fixed yet though prior this can be merged. Explicitly:
* there's a conflict for ``` shared/transforms/combinechecks.py``` right now -- the ```checks()``` routine has been introduced in some of the recent commits. Fix that (adding product in ```combinechecks.py``` should be now handled within the ```checks()``` routine) -- add your changes to that part && increment count of position arguments in the corresponding for loop,
* as suggested by Gabe move the version back to 5.10 in:
  * ```shared/modules/testcheck_module.py```,
  * ```shared/utils/verify-input-sanity.py```,

We can temporarily tolerate absence of proper ```testcheck_module.py``` and ```verify-input-sanity.py``` scripts for Fedora (in favour of keeping properly working RHEL content). It's not a blocker for merge. Though if you want to have a look a way should be investigated how to make ```testcheck_module.py``` && ```verify-input-sanity.py``` scripts to understand / accept different OVAL language versions in their ```header()``` routines (IOW make the ```header()``` routines there to accept OVAL language version as argument). But this can be done in some of future PRs.

* the former ```multi_platform_rhel``` occurrences seem to have been fixed already (thanks for that!),
* ad calling custom ```./compose_ds.sh``` and ```./validate_sh``` scripts in Fedora Makefile -- I would propose to get rid of them and use ```1.2.2``` (the actually required openscap-scanner version respectively) in the ```BuildRequires``` field of the ```Fedora``` section of the ```scap-security-guide.spec.in``` SPEC file. As already pointed out by Gabe, openscap-1.2.2 has reached stable channel version in Fedora already.

 When building the content on RHEL, we don't include the Fedora content there (the official / native Red Hat rpms aren't doing that, and assuming we shouldn't be doing that too not to cause confusion). The ```common``` profile in Fedora is provided just for administrator's convenience to make the Fedora system hardened via automated ways, not with intention underlying Fedora system to be used as a certified system. So there's isn't a need (right now and also in the future) the Fedora content to be shipped when RPM package is built on Red Hat Enterprise Linux system. For users despite of the above paragraph clarification still interested in Fedora content on Red Hat Enterprise Linux, we can create a blog post / wiki guide how to enable it.

Considering the conclusion from previous paragraph custom ```compose_ds.sh``` and ```validate_sh``` shouldn't be necessary. Use Fedora specific ```BuildRequires / Requires``` instead of them.

Thanks, Jan.

jan-cerny commented:
* Resolved conflict of combinechecks.py
* Set OVAL back to 5.10
* Removed occurences of multi_platform_rhel
* Removed wrapper scripts in Makefile, updated Makefile
(I haven't modified the specfile yet)

jan-cerny commented:
Modified a specfile. Build requires OpenSCAP 1.2.2 when building in Fedora.

There is an issue with building, I had to comment out occurrences of `scap-security-guide.8`, this need to be fixed in other PR

iankko commented:
@jan-cerny 
> Modified a specfile. Build requires OpenSCAP 1.2.2 when building in Fedora.

Thanks, Jan. This looks better!

> There is an issue with building, I had to comment out occurrences of scap-security-guide.8, this need to be fixed in other PR

Yeah, it's a known issue (for like ~2-3 weeks) ```make rpm``` fails on RHEL && Fedora (due to unfinished ```make install`` changes yet).

We will fix this in some future PR, no worries (I will create a PR for that).

Thank you. ACK && Merging this.
